elastic-buffer-sim/exe/Main.hs

```diff
@@ -2,70 +2,644 @@
+  writeFile topologyFile $ (<> "\n") $ render $ toDot g name
+  BS.writeFile (outDir </> "simulate.json") $ encode options
+     { stable   = isStable
+     , offsets  = offs
+     , jsonArgs = Nothing
+     , topology = case topology of
+         Just (Random _) -> Just $ DotFile topologyFile
+         _               -> topology
+     }
+
+  if isStable
+  then exitSuccess
+  else exitFailure
+
+randomGraph :: Int -> IO Graph
+randomGraph n = do
+  let is = [0,1..n - 1]
+
+  -- get some random vertex permuation for ensuring connectivity
+  aP <- newListArray (0, n - 1) is
+  replicateM_ (10 * n) $ do
+    p1 <- randomRIO (0, n - 1)
+    p2 <- randomRIO (0, n - 1)
+    v1 <- readArray aP p1
+    v2 <- readArray aP p2
+    writeArray aP p1 v2
+    writeArray aP p2 v1
+
+  randomPermutation <- getElems (aP :: IOUArray Int Int)
+
+  -- get some random edge availability matrix for self-loop-free,
+  -- undirected graphs
+  aE <- replicateM (n * n) randomIO >>= newListArray ((0, 0), (n - 1, n - 1))
+  forM_ is $ \i ->
+    writeArray aE (i, i) False
+  forM_ [(i, j) | i <- is, j <- is, i /= j] $ \(i, j) -> do
+    d1 <- readArray aE (i, j)
+    d2 <- readArray aE (j, i)
+    when (d1 /= d2) $ do
+      writeArray aE (i, j) False
+      writeArray aE (j, i) False
+
+  -- ensure connectivity
+  forM_ (unchain randomPermutation) $ \(i, j) -> do
+    writeArray aE (i, j) True
+    writeArray aE (j, i) True
+
+  available <- freeze (aE :: IOUArray (Int, Int) Bool)
+
+  -- create the graph
+  return $ buildG (0, n - 1)
+    [ (i, j)
+    | i <- is
+    , j <- is
+    , available A.! (i, j)
+    ]
+
+toDot :: Graph -> String -> (Bool, GraphType, Maybe Name, [Statement])
+toDot g name = (True, Graph, Just $ StringID name, map asEdgeStatement $ edges g)
+ where
+  asEdgeStatement (x, y) =
+    EdgeStatement (map (\i -> NodeRef (XMLID ('n' : show i)) Nothing) [x,y]) []
+
+fromDot :: String -> Either String (Graph, String)
+fromDot cnt = do
+  (strict, gType, maybe "" fromDotName -> name, statements) <- parse cnt
+
+  when (not strict) $ Left "Graph must be strict"
+  when (gType == Digraph) $ Left "Graph must be undirected"
+
+  namedEdges <-
+    fmap
+      ( concatMap (map (\(x,y) -> (asString x, asString y)) . unchain)
+      . catMaybes
+      )
+    $ forM statements $ \case
+        EdgeStatement xs _ -> fmap Just $ forM xs $ \case
+          NodeRef n _ -> return n
+          Subgraph{}  -> Left "Subgraphs are not supported"
+        _ -> pure Nothing
+
+  let
+    edgeNames = strictSort $ concatMap (\(x, y) -> [x, y]) namedEdges
+    n = length edgeNames - 1
+    idx = (M.!) $ M.fromList $ zip edgeNames [0,1..]
+    g = buildG (0,n)
+      $ filter (uncurry (/=))
+      $ strictSort
+      $ concatMap (\(x, y) -> [(idx x, idx y), (idx y, idx x)])
+        namedEdges
+
+  when (length (scc g) > 1) $ Left "Graph must be strongly connected"
+
+  return (g, name)
+
+ where
+  strictSort :: Ord a => [a] -> [a]
+  strictSort = S.toList . S.fromList
```

Comment:
```suggestion
  strictSort :: Ord a => [a] -> [a]
  strictSort = S.toList . S.fromList
```

This doesn't _just_ sort, it also de-duplicates. This should be reflected in the name. (Do we want to de-duplicate? If so, it's probably clearer to just use a separate `nubOrd` step. This is more expensive, but this code isn't really performance sensitive. Code clarity >> performance for non-performance critical code.)

Reply:
> This doesn't just sort, it also de-duplicates.

This is what the `strict` in front of `sort` is meant to indicate. `strictSort` was a very common term during my FP courses, so it feels very natural to me. However, just searching for it revealed that I maybe got a little bit biased from my university context there.

>Do we want to de-duplicate?

Technically, we don't. The dot specification disallows duplicate edges when using the `strict` keyword, which is required for topologies. On the other hand, it is user input, thus I never trust it and the dot reading library does not check for it either. Furthermore, it's totally legal to delete duplicates form the graph at application side, given to the `stirct` keyword. So why not, it doesn't hurt at least.

> If so, it's probably clearer to just use a separate nubOrd step. ... Code clarity >> performance for non-performance critical code.

I actually just had to look up `nubOrd` first to get what it is doing. Therefore, `S.toList . S.fromList` is much clearer in the that sense to me, it has the same computational cost and even gets the list sorted, which improves readability of your data. Sets also enjoy longer support than `nubOrd` in `containers`. So that's the clear winner for me at this point ;-).

Reply:
Alright, that makes sense. That puts it in the list of "strictly ascending". However, as you can see, there is ambiguity here: "strict" usually refers to the-opposite-of-lazy, which is the way I read it. I didn't consider it could mean "deduplicate and sort". Which brings me too:

> I actually just had to look up `nubOrd` first to get what it is doing.

This type of confusion is fine (within reason of course, no need to drag in category theory nomenclature everywhere). It's the leading astray type of confusion we should strive to prevent. So let's remove the ambiguity by calling it something along the lines of:

```haskell
nubAndSort   -- combines 'nub' and 'sort' from 'Prelude'
dedupAndSort -- replaces 'nub' by a more sensible word
```

> Sets also enjoy longer support than nubOrd in containers.

I'm assuming you're implying that this therefore has a higher chance of being recognized by Haskell engineers. I don't think so; `nub` has been part of the Haskell98 report. `nubOrd` is its better performing cousin.

Personally would write it as:

```
 $ sort   -- for pretty printing
 $ nubOrd -- remove duplicate edges
 $ ..
```

but

```
  $ dedupAndSort -- remove duplicate edges, sort for pretty printing
``` 

is fine too. In any case, explain the choices, I wasn't sure why things were the way the were.

Reply:
Done.